features/dd-sdk-android-trace/src/test/kotlin/com/datadog/opentracing/SpanCleanerTest.kt

```diff
@@ -0,0 +1,184 @@
+
+    private lateinit var testeCleaner: SpanCleaner
+
+    @BeforeEach
+    fun setUp() {
+        testeCleaner = SpanCleaner()
+    }
+
+    // region cleanup
+
+    @Test
+    fun `M remove random traces W add and remove`(forge: Forge) {
+        // Given
+        val pendingTraces = forge.aList(size = forge.anInt(min = 10, max = 50)) { mock<PendingTrace>() }
+        pendingTraces.forEach {
+            testeCleaner.addPendingTrace(it)
+        }
+        assertThat(testeCleaner.pendingTraces.values).containsExactlyInAnyOrderElementsOf(pendingTraces)
+
+        // When
+        pendingTraces.shuffled().forEach {
+            testeCleaner.removePendingTrace(it)
+        }
+
+        // Then
+        assertThat(testeCleaner.pendingTraces).isEmpty()
+    }
+
+    @Test
+    fun `M cleanup random traces W add and close`(forge: Forge) {
+        // Given
+        val pendingTraces = forge.aList(size = forge.anInt(min = 10, max = 50)) { mock<PendingTrace>() }
+        pendingTraces.forEach {
+            testeCleaner.addPendingTrace(it)
+        }
+        assertThat(testeCleaner.pendingTraces.values).containsExactlyInAnyOrderElementsOf(pendingTraces)
+
+        // When
+        testeCleaner.close()
+
+        // Then
+        pendingTraces.forEach {
+            verify(it).clean()
+        }
+    }
+
+    // endregion
+
+    // region `equals` contract
+
+    @Test
+    fun `M contain only one entry W add same PendingTrace twice`(forge: Forge) {
+        // Given
+        val pendingTrace = mockPendingTrace(forge)
+        testeCleaner.addPendingTrace(pendingTrace)
+        testeCleaner.addPendingTrace(pendingTrace)
+
+        // Then
+        assertThat(testeCleaner.pendingTraces).hasSize(1)
+        assertThat(testeCleaner.pendingTraces.values).containsExactly(pendingTrace)
+    }
+
+    @Test
+    fun `M contain only one entry W add {same PendingTrace mutated, twice}`(forge: Forge) {
+        // Given
+        val pendingTrace = mockPendingTrace(forge)
+        testeCleaner.addPendingTrace(pendingTrace)
+        pendingTrace.addSpan(mock())
+        pendingTrace.addSpan(mock())
+        testeCleaner.addPendingTrace(pendingTrace)
+
+        // Then
+        assertThat(testeCleaner.pendingTraces).hasSize(1)
+        assertThat(testeCleaner.pendingTraces.values).containsExactly(pendingTrace)
+
+        // When
+        testeCleaner.removePendingTrace(pendingTrace)
+
+        // Then
+        assertThat(testeCleaner.pendingTraces).isEmpty()
+    }
+
+    // endregion
+
+    // region concurrency
+
+    @Test
+    fun `M not leak any PendingTrace W add, mutate and remove { different threads }`(forge: Forge) {
+        // Given
+        val pendingTraces = forge.aList(size = forge.anInt(min = 10, max = 50)) { mockPendingTrace(forge) }
+        val addCountDownLatch = CountDownLatch(pendingTraces.size)
+        val removeCountDownLatch = CountDownLatch(pendingTraces.size)
+        pendingTraces.take(pendingTraces.size).forEach { pendingTrace ->
```

Comment:
I'm confused by the `pendingTraces.take(pendingTraces.size)` -  it is essentially the same as without `take` call, no? It is effectively the same collection. Should it be `pendingTraces.take(pendingTraces.size / 2)` here?